### PR TITLE
[api] Impletment TestGrpcServer_GetActionsByAddress by gomock CoreSer…

### LINKED ISSUE #3223: [api] impletment TestGrpcServer_GetActionsByAddress by gomock CoreService interface
**What would you like to be added**:
impletment `TestGrpcServer_GetActionsByAddress` in `./api/grpcserver_test.go`  by  `gomock` [CoreService interface](https://github.com/iotexproject/iotex-core/blob/master/api/coreservice.go#L61)

see exp:  [TestGrpcServer_SendAction](https://github.com/iotexproject/iotex-core/blob/master/api/grpcserver_test.go#L58)

**Why is this needed**:

### PR DESCRIPTION
**Pull request by LuckyPigeon**
# Description

impletment `TestGrpcServer_GetActionsByAddress` in `./api/grpcserver_test.go` by gomock [CoreService interface](https://github.com/iotexproject/iotex-core/blob/master/api/coreservice.go#L61)

Fixes #3223 

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [x] TestGrpcServer_GetActionsByAddress

**Test Configuration**:
- Firmware version: Ubuntu 21.04
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/grpcserver_test.go [hunk 1]
```diff
@@ -36,7 +36,38 @@ func TestGrpcServer_GetAction(t *testing.T) {
...
+	grpcSvr := NewGRPCServer(core, testutil.RandomPort())
 
+	for _, test := range _getActionsByAddressTests {
+		actInfo := &iotexapi.ActionInfo{
+			Index:   0,
+			ActHash: "test",
+		}
+		response := []*iotexapi.ActionInfo{}
+		request := &iotexapi.GetActionsRequest{
+			Lookup: &iotexapi.GetActionsRequest_ByAddr{
```
[the remark this answers is not in the document]
Comment (LuckyPigeon) [reply api/grpcserver_test.go:52]: Should we combine this PR and above types of request into #3221, and remove this test and `TestGrpcServer_GetActionsByBlock`?

**Comment by LuckyPigeon** [comment]
ci test seems can't resolve `testutil`.

File: api/grpcserver_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (LuckyPigeon) [reply api/grpcserver_test.go:51]: I'll do this in #3221

[the remark this answers is not in the document]
Comment (huof6829) [reply api/grpcserver_test.go]: I combine types in one function. I think it's more clear.

File: api/grpcserver_test.go [hunk 2]
```diff
@@ -35,11 +35,98 @@ func TestGrpcServer_GetAction(t *testing.T) {
...
+		for _, request := range requests {
+			request.call()
+			result, err := grpcSvr.GetActions(context.Background(), request.req)
+			require.NoError(err)
+			require.Equal(uint64(test.numActions), result.Total)
+		}
+	}
 }
 
 func TestGrpcServer_GetUnconfirmedActionsByAddress(t *testing.T) {
```
Comment (huof6829) [inline api/grpcserver_test.go:97]: delete

File: api/grpcserver_test.go [hunk 3]
```diff
@@ -35,11 +35,98 @@ func TestGrpcServer_GetAction(t *testing.T) {
 }
 
 func TestGrpcServer_GetActionsByAddress(t *testing.T) {
```
Comment (huof6829) [inline api/grpcserver_test.go:37]: TestGrpcServer_GetActions

File: api/grpcserver_test.go [hunk 4]
```diff
@@ -35,11 +35,98 @@ func TestGrpcServer_GetAction(t *testing.T) {
...
+							Address: test.address,
+							Start:   test.start,
+							Count:   test.count,
+						},
+					},
+				},
+				call: func() {
+					core.EXPECT().UnconfirmedActionsByAddress(gomock.Any(), gomock.Any(), gomock.Any()).Return(response, nil)
+				},
+			},
```
Comment (huof6829) [inline api/grpcserver_test.go:85]: you can try to add other types like this.

File: api/grpcserver_test.go [hunk 5]
```diff
@@ -26,20 +26,142 @@ func TestGrpcServer_GetAccount(t *testing.T) {
 
 }
 
-func TestGrpcServer_GetActions(t *testing.T) {
-
-}
-
 func TestGrpcServer_GetAction(t *testing.T) {
```
Comment (huof6829) [inline api/grpcserver_test.go:29]: delete

File: api/grpcserver_test.go [hunk 6]
```diff
@@ -26,20 +26,142 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+		response := []*iotexapi.ActionInfo{}
+		request := &iotexapi.GetActionsRequest{
+			Lookup: &iotexapi.GetActionsRequest_UnconfirmedByAddr{
+				UnconfirmedByAddr: &iotexapi.GetUnconfirmedActionsByAddressRequest{
+					Address: test.address,
+					Start:   test.start,
+					Count:   test.count,
+				},
+			},
+		}
```
Comment (huof6829) [inline api/grpcserver_test.go:153]: similar with line 71~75 ?  if yes, delete.  if no, combine them

File: api/grpcserver_test.go [hunk 7]
```diff
@@ -26,20 +26,142 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+
+			for _, request := range requests {
+				request.call()
+				result, err := grpcSvr.GetActions(context.Background(), request.req)
+				require.NoError(err)
+				require.Equal(uint64(test.numActions), result.Total)
+			}
+		}
+	})
+
```
Comment (huof6829) [inline api/grpcserver_test.go:106]: add  type `iotexapi.GetActionsRequest_ByBlk` to check

**Comment by LuckyPigeon** [comment]
@huof6829 `block.Block.Actions` is unabled to access, so they sub function of `actionsInBlock` can't be full tested.

**Comment by huof6829** [comment]
> @huof6829 `block.Block.Actions` is unabled to access, so they sub function of `actionsInBlock` can't be full tested.

Because block.Actions is not add in test line157,  so  `actionsInBlock` returns https://github.com/iotexproject/iotex-core/blob/0d72eb73842e17be6651169b4b8e05b4a2648504/api/grpcserver.go#L245-L246

File: api/grpcserver_test.go [hunk 8]
```diff
@@ -47,23 +50,205 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+					},
+				},
+				Receipts: []*action.Receipt{
+					{
+						BlockHeight: test.blkHeight,
+						GasConsumed: gasConsumed.Uint64(),
+					},
+				},
+			}
+			fmt.Println(test.start)
```
Comment (huof6829) [inline api/grpcserver_test.go:179]: delete

File: api/grpcserver_test.go [hunk 9]
```diff
@@ -47,23 +50,205 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+			core.EXPECT().Action(gomock.Any(), gomock.Any()).Return(response, nil)
+
+			result, err := grpcSvr.GetActions(context.Background(), request)
+			require.NoError(err)
+			require.Equal(test.blkNumber, result.ActionInfo[0].BlkHeight)
+		}
+	})
 
-}
+	t.Run("get actions by block test", func(t *testing.T) {
```
Comment (huof6829) [inline api/grpcserver_test.go:150]: is it same with line208~250?

File: api/grpcserver_test.go [hunk 10]
```diff
@@ -47,23 +50,205 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+						BlockHeight: test.blkHeight,
+						GasConsumed: gasConsumed.Uint64(),
+					},
+				},
+			}
+			fmt.Println(test.start)
+			// for i := 0; i < int(test.start); i++ {
+			// 	response.Block.Actions = append(response.Block.Actions, tsf1)
+			// }
+			response.Block.Actions = []action.SealedEnvelope{tsf1, tsf1}
```
Comment (huof6829) [inline api/grpcserver_test.go:183]: no need to repeat tsf1

File: api/grpcserver_test.go [hunk 11]
```diff
@@ -47,23 +49,147 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+						CheckPending: test.checkPending,
+					},
+				},
+			}
+
+			core.EXPECT().Action(gomock.Any(), gomock.Any()).Return(response, nil)
+
+			result, err := grpcSvr.GetActions(context.Background(), request)
+			require.NoError(err)
+			require.Equal(test.blkNumber, result.ActionInfo[0].BlkHeight)
```
Comment (huof6829) [inline api/grpcserver_test.go:145]: check  len(result.ActionInfo)  before using  result.ActionInfo[0]

File: api/grpcserver_test.go [hunk 12]
```diff
@@ -47,23 +49,147 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+			}
+
+			core.EXPECT().BlockByHash(gomock.Any()).Return(response, nil).AnyTimes()
+
+			_, err := grpcSvr.GetActions(context.Background(), request)
+			if test.count == 0 {
+				require.Error(err)
+				continue
+			}
+			require.NoError(err)
```
Comment (huof6829) [inline api/grpcserver_test.go:190]: can check result with  `numActions`  like line120 ?

File: api/grpcserver_test.go [hunk 13]
```diff
@@ -47,23 +49,147 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+						Count:   test.count,
+					},
+				},
+			}
+
+			core.EXPECT().BlockByHash(gomock.Any()).Return(response, nil).AnyTimes()
+
+			_, err := grpcSvr.GetActions(context.Background(), request)
+			if test.count == 0 {
+				require.Error(err)
```
Comment (huof6829) [inline api/grpcserver_test.go:187]: can you check err content?

**Review by huof6829 — APPROVED** [review]

**Comment by LuckyPigeon** [comment]
@huof6829 Thanks for the improvement!

File: api/grpcserver_test.go [hunk 14]
```diff
@@ -47,23 +49,146 @@ func TestGrpcServer_GetAccount(t *testing.T) {
...
+			for _, request := range requests {
+				request.call()
+				result, err := grpcSvr.GetActions(context.Background(), request.req)
+				require.NoError(err)
+				require.Equal(uint64(test.numActions), result.Total)
+			}
+		}
+	})
+
+	t.Run("get actions by action test", func(t *testing.T) {
```
Comment (Liuhaai) [inline api/grpcserver_test.go:125]: get actions by action?

Comment (LuckyPigeon) [reply api/grpcserver_test.go:125]: There's a case where method named `svr.coreService.Action`, it purpose is to get a single action. But I can see that the naming isn't perfect, will configure a proper name.

Comment (CoderZhi) [reply api/grpcserver_test.go:125]: get actions by hash

**Review by CoderZhi — APPROVED** [review]

File: api/grpcserver_test.go (hunk 14 above)
Comment (LuckyPigeon) [reply api/grpcserver_test.go]: Done!

**Pull request merged**